### DISCUSSION TITLE
Infinispan TLS Implementation

### DISCUSSION THREAD

**Discussion by tastyfrankfurt** [keycloak/keycloak#11761] [Q&A]
Hi Team,

I have run into this bug, https://github.com/keycloak/keycloak/issues/11645, but on top of that i am unable to configure TLS in infinispan as it seems to be missing the libraries required,  the below is a configuration snippet to enable TLS but it gets rejected as the java name space is missing,

`<security xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="urn:infinispan:server:13.0
        https://infinispan.org/schemas/infinispan-server-13.0.xsd"
        xmlns="urn:infinispan:server:13.0">
    <security-realms>
        <security-realm name="default">
            <server-identities>
                <ssl>
                    <keystore path="keystore.jks"
                        relative-to="infinispan.server.config.path"
                        keystore-password="REDACTED"
                        alias="localhost"/>
                    <engine enabled-protocols="TLSv1.2"
                        enabled-ciphersuites="SSL_RSA_WITH_AES_256_GCM_SHA256"/>
                </ssl>
            </server-identities>
            <truststore-realm path="trust.jks"
                    relative-to="infinispan.server.config.path"
                    keystore-password="REDACTED"/>
        </security-realm>
    </security-realms>
</security>`

2022-04-29 06:05:27,334 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (production) mode
2022-04-29 06:05:27,334 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start caches
2022-04-29 06:05:27,334 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: org.infinispan.commons.CacheConfigurationException: ISPN000327: Cannot find a parser for element 'security' in namespace 'urn:infinispan:server:13.0'. Check that your configuration is up-to date for Infinispan '13.0.8.Final' and if you have the proper dependency in the classpath
2022-04-29 06:05:27,334 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: ISPN000327: Cannot find a parser for element 'security' in namespace 'urn:infinispan:server:13.0'. Check that your configuration is up-to date for Infinispan '13.0.8.Final' and if you have the proper dependency in the classpath
2022-04-29 06:05:27,334 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) For more details run the same command passing the '--verbose' option. Also you can use '--help' to see the details about the usage of the particular command.

Is someone able, to point me at an existing feature request or bug for this too please?  This is a compliance requirement for our project.

Thanks

TastyFrankfurt

**Comment by AlexeiKlimenko** [comment]
Have you resolved this issue?

**Reply by tastyfrankfurt** [reply]
No issue is still outstanding, we have moved onto using an alternative authentication platform due to the security issues with trying to setup a cluster in keycloak.  If we had a servicemesh layer in a k8s environment then we could solve this externally to the application, but having nodes communicate authentication information in clear text was causing us issues with meeting our industry compliance requirements.

Thanks

Tasty

**Comment by tastyfrankfurt** [comment]
Working on version 22.0.1 this is still an issue

**Reply by tastyfrankfurt** [reply]
Okay found the answer, for anyone deploying keycloak and wanting to use tcp and tls security for the infinispan server here is an example configuration for use, note you do have to do a kc.sh build to get it injected, and create cert jks store and a truststore, TCP ports 7800 and 2157(undocumented).  Good luck and good hunting everyone.


```
<?xml version="1.0" encoding="UTF-8"?>
<infinispan xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="urn:infinispan:config:14.0 http://www.infinispan.org/schemas/infinispan-config-14.0.xsd"
        xmlns="urn:infinispan:config:14.0">
    <jgroups>
        <stack name="encrypt-tcp">
            <TCP bind_addr="<hostname>" 
                    bind_port="7800" 
                    diag.enabled="${jgroups.diag.enabled:false}"
                    thread_naming_pattern="pl"
                    send_buf_size="640k"
                    sock_conn_timeout="300"
                    bundler_type="transfer-queue"
                    bundler.max_size="${jgroups.bundler.max_size:64000}"
                    thread_pool.min_threads="${jgroups.thread_pool.min_threads:0}"
                    thread_pool.max_threads="${jgroups.thread_pool.max_threads:200}"
                    thread_pool.keep_alive_time="60000"
                    thread_pool.thread_dumps_threshold="${jgroups.thread_dumps_threshold:10000}"/>
            <TCPPING initial_hosts="hosta[7800],hostb[7800]"
                    port_range="2"/>
            <MERGE3 min_interval="10000"
                    max_interval="30000"/>
            <FD_ALL timeout="8000" interval="3000"/>
            <FD_SOCK />
            <FD_ALL3/>
            <VERIFY_SUSPECT timeout="1000"/>
            <SSL_KEY_EXCHANGE keystore_name="/opt/keycloak/conf/certs/keycloak.jks"
                            keystore_password="changeme"
                            truststore_name="/opt/keycloak/conf/certs/keycloaktruststore.jks"
                            truststore_password="changeme"/>
            <ASYM_ENCRYPT asym_keylength="4096"
                        asym_algorithm="RSA"
                        change_key_on_coord_leave = "false"
                        change_key_on_leave = "false"
                        use_external_key_exchange = "true"/>
            <pbcast.NAKACK2 use_mcast_xmit="false"
                        xmit_interval="200"
                        xmit_table_num_rows="50"
                        xmit_table_msgs_per_row="1024"
                        xmit_table_max_compaction_time="30000"
                        resend_last_seqno="true"/>
            <UNICAST3 conn_close_timeout="5000"
                        xmit_interval="200"
                        xmit_table_num_rows="50"
                        xmit_table_msgs_per_row="1024"
                        xmit_table_max_compaction_time="30000"/>
            <pbcast.STABLE desired_avg_gossip="5000"
                        max_bytes="1M"/>
            <pbcast.GMS print_local_addr="false"
                        join_timeout="${jgroups.join_timeout:2000}" />
            <UFC max_credits="${jgroups.max_credits:4m}"
                        min_threshold="0.40"/>
            <MFC max_credits="${jgroups.max_credits:4m}"
                        min_threshold="0.40"/>
            <FRAG4 frag_size="${jgroups.frag_size:60000}"/>
        </stack>
    </jgroups>
    <cache-container name="keycloak" statistics="true">
        <transport stack="encrypt-tcp" node-name="<hostname>"/>
        <local-cache name="realms" simple-cache="true">
            <encoding>
                <key media-type="application/x-java-object"/>
                <value media-type="application/x-java-object"/>
            </encoding>
            <memory max-count="10000"/>
        </local-cache>
        <local-cache name="users" simple-cache="true">
            <encoding>
                <key media-type="application/x-java-object"/>
                <value media-type="application/x-java-object"/>
            </encoding>
            <memory max-count="10000"/>
        </local-cache>
        <distributed-cache name="sessions" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <distributed-cache name="authenticationSessions" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <distributed-cache name="offlineSessions" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <distributed-cache name="clientSessions" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <distributed-cache name="offlineClientSessions" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <distributed-cache name="loginFailures" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <local-cache name="authorization" simple-cache="true">
            <encoding>
                <key media-type="application/x-java-object"/>
                <value media-type="application/x-java-object"/>
            </encoding>
            <memory max-count="10000"/>
        </local-cache>
        <replicated-cache name="work">
            <expiration lifespan="-1"/>
        </replicated-cache>
        <local-cache name="keys" simple-cache="true">
            <encoding>
                <key media-type="application/x-java-object"/>
                <value media-type="application/x-java-object"/>
            </encoding>
            <expiration max-idle="3600000"/>
            <memory max-count="1000"/>
        </local-cache>
        <distributed-cache name="actionTokens" owners="2">
            <encoding>
                <key media-type="application/x-java-object"/>
                <value media-type="application/x-java-object"/>
            </encoding>
            <expiration max-idle="-1" lifespan="-1" interval="300000"/>
            <memory max-count="-1"/>
        </distributed-cache>
    </cache-container>
</infinispan>
```
